### PR TITLE
Add Focus support for iOS platform view

### PR DESCRIPTION
**Pull request by hellohuanlin**
This PR makes iOS platform view focusable. 

Since frameworks and engines are separate repos, I have to try-and-catch the non-existing "viewFocused" channel message for now. Will remove the try-catch after we landed and rolled the engine update. 

The main idea is that: 

1. In the framework, a platform view needs to be wrapped under FocusNode with a `onFocusChange` callback saved
2. In the engine, when a platform view is focused, it should send a channel message "viewFocused" to framework, which will invoke the `onFocusChange` saved in the previous step
3. The `onFocusChange` will then send a channel message "TextInput.setPlatformViewClient" to the engine, which should remove the focus on previously focused text inputs (e.g. EditableText)

I am going through the Flutter Style Guide now (will check the box once I am done). But I tried my best to following existing conventions.

## Related Issue

https://github.com/flutter/flutter/issues/34187

## Pre-launch Checklist

I am going through the Flutter Style Guide now (will check the box once I am done). But I tried my best to following existing conventions. 

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 1]
```diff
@@ -584,6 +590,7 @@ class _UiKitViewState extends State<UiKitView> {
     }
     _initialized = true;
     _createNewUiKitView();
+    _focusNode = FocusNode(debugLabel: "UiKitView(id: $_id)");
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/platform_view.dart:593]: You wouldn't need to set this if you declared it as `  late final FocusNode _focusNode = FocusNode(debugLabel: 'UiKitView(id: $_id)');`
```suggestion
```
But really, is this the right place to set this?  What if the `_id` changes?  It seems like it should be initialized in `setState`?
Also this should be single-quotes.

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 2]
```diff
@@ -639,13 +646,37 @@ class _UiKitViewState extends State<UiKitView> {
...
+    if (!isFocused) {
+      // Unlike Android, we do not need to send "clearFocus" channel message
+      // to the engine, because focusing on another view will automatically
+      // cancel the focus on the previously focused platform view.
+    } else {
+      SystemChannels.textInput.invokeMethod<void>(
+        'TextInput.setPlatformViewClient',
+        <String, dynamic>{'platformViewId': _id},
+      ).catchError((dynamic e) {
+        // TODO(huan): remove this once "TextInput.setPlatformViewClient" is implemented on engine.
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/platform_view.dart:674]: Can you instead land this after the engine change?

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 3]
```diff
@@ -2039,7 +2092,14 @@ void main() {
       // is not yet in the tree.
       await tester.pump();
 
-      final SemanticsNode semantics = tester.getSemantics(find.byType(UiKitView));
+      final SemanticsNode semantics = tester.getSemantics(
+        find.descendant(
+          of: find.byType(UiKitView),
+          matching: find.byWidgetPredicate(
+              (Widget widget) => widget.runtimeType.toString() == "_UiKitPlatformView",
```
Comment (jmagman) [inline packages/flutter/test/widgets/platform_view_test.dart:2099]: Single quotes
```suggestion
              (Widget widget) => widget.runtimeType.toString() == '_UiKitPlatformView',
```

File: packages/flutter/lib/src/services/platform_views.dart [hunk 4]
```diff
@@ -227,7 +230,9 @@ class PlatformViewsService {
       );
     }
     await SystemChannels.platform_views.invokeMethod<void>('create', args);
-    return UiKitViewController._(id, layoutDirection);
+    final UiKitViewController controller = UiKitViewController._(id, layoutDirection);
+    _instance._focusCallbacks[id] = onFocus ?? () {};
```
Comment (jmagman) [inline packages/flutter/lib/src/services/platform_views.dart:234]: How about only adding if it's non-null instead of an empty block?
```suggestion
    if (onFocus != null) {
      _instance._focusCallbacks[id] = onFocus;
    }
```

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 5]
```diff
@@ -639,13 +646,37 @@ class _UiKitViewState extends State<UiKitView> {
...
     );
     if (!mounted) {
       controller.dispose();
       return;
     }
     widget.onPlatformViewCreated?.call(id);
-    setState(() { _controller = controller; });
+    setState(() {
+      _controller = controller;
+      _id = id;
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/platform_view.dart:660]: Per previous comment, does the `focusNode` `debugLabel` need to be updated when `_id` changes?

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 6]
```diff
@@ -559,9 +559,11 @@ class _AndroidViewState extends State<AndroidView> {
 }
 
 class _UiKitViewState extends State<UiKitView> {
+  int? _id;
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/platform_view.dart:562]: Can you instead use `_controller.id` where needed?

**Review by jmagman — COMMENTED** [review]
Some nits, @cyanglaz knows more about whether the actual focus logic is correct.

**Comment by jmagman** [comment]
Also the ci.yaml failure is because you're branched off a commit from a few weeks ago, it's been fixed if you rebase to top of tree.

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 2 above)
Comment (hellohuanlin) [reply packages/flutter/lib/src/widgets/platform_view.dart:674]: If we land engine first, there would be a similar problem when the engine sends "focusView" channel message that's not implemented in framework yet.

**Comment by jmagman** [comment]
Friendly ping to take a look at this one @cyanglaz

File: packages/flutter/lib/src/services/platform_views.dart [hunk 7]
```diff
@@ -227,7 +230,11 @@ class PlatformViewsService {
       );
     }
     await SystemChannels.platform_views.invokeMethod<void>('create', args);
-    return UiKitViewController._(id, layoutDirection);
+    final UiKitViewController controller = UiKitViewController._(id, layoutDirection);
```
Comment (cyanglaz) [inline packages/flutter/lib/src/services/platform_views.dart:233]: nits:
I don't see the controller variable being used. Maybe keep the original code about returning the controller so the diff is smaller?

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 8]
```diff
@@ -562,6 +562,7 @@ class _UiKitViewState extends State<UiKitView> {
   UiKitViewController? _controller;
   TextDirection? _layoutDirection;
   bool _initialized = false;
+  FocusNode? _focusNode;
```
Comment (cyanglaz) [inline packages/flutter/lib/src/widgets/platform_view.dart:565]: Is there a situation where _focusNode is nil and needs to be accessed?
Maybe `late FocusNode _focusNode;` is better here?

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 2 above)
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart:672]: Maybe split the PRs a little more so the landing orders would be:
1. Framework receive method channel (it will be a non-op after landing it as the method channel will never be fired from engine)
1. Engine implementation.
1. Framework send method channel 

This is a little more complicated but we can ensure there is no bad version or TODOs like this.

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 9]
```diff
@@ -1978,7 +1978,7 @@ void main() {
       },
     );
 
-    testWidgets('AndroidView rebuilt with same gestureRecognizers', (WidgetTester tester) async {
+    testWidgets('UiKitView rebuilt with same gestureRecognizers', (WidgetTester tester) async {
```
Comment (cyanglaz) [inline packages/flutter/test/widgets/platform_view_test.dart:1981]: thank you!!!

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 10]
```diff
@@ -2039,7 +2092,14 @@ void main() {
...
       await tester.pump();
 
-      final SemanticsNode semantics = tester.getSemantics(find.byType(UiKitView));
+      final SemanticsNode semantics = tester.getSemantics(
+        find.descendant(
+          of: find.byType(UiKitView),
+          matching: find.byWidgetPredicate(
+              (Widget widget) => widget.runtimeType.toString() == '_UiKitPlatformView',
+          ),
+        ),
```
Comment (cyanglaz) [inline packages/flutter/test/widgets/platform_view_test.dart:2101]: Why is this change needed? Does adding a focus node changing how semantics nodes are created?

Comment (hellohuanlin) [reply packages/flutter/test/widgets/platform_view_test.dart:2101]: Yes, when we wrap it under a `Focus` node, the original `_UiKitPlatformView` got nested one level deeper in the tree

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 2 above)
Comment (hellohuanlin) [reply packages/flutter/lib/src/widgets/platform_view.dart:672]: @cyanglaz I think i can do that.

File: packages/flutter/test/widgets/platform_view_test.dart (hunk 10 above)
Comment (cyanglaz) [reply packages/flutter/test/widgets/platform_view_test.dart:2101]: oh I see `find.byType` is not recursive.

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 11]
```diff
@@ -639,13 +644,23 @@ class _UiKitViewState extends State<UiKitView> {
...
     widget.onPlatformViewCreated?.call(id);
-    setState(() { _controller = controller; });
+    setState(() {
+      _controller = controller;
+      _focusNode = FocusNode(debugLabel: 'UiKitView(id: $id)');
+    });
+  }
+
+  void _onFocusChange(bool isFocused) {
+    // TODO: (hellohuanlin) send 'TextInput.setPlatformViewClient' channel message to engine after the engine is updated to handle this message.
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/platform_view.dart:663]:
```
   info â�¢ Use Flutter TODO format: // TODO(username): message, https://URL-to-issue â�¢ packages/flutter/lib/src/widgets/platform_view.dart:663:5 â�¢ flutter_style_todos
```

**Comment by hellohuanlin** [comment]
@cyanglaz can you take another look? thx

**Review by cyanglaz — APPROVED** [review]
LGTM
